src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +73,200 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
+// The terminal needs to be configured different depending on the user.
+// ConfigureTerminalForXXX are called to change the configuration.
+// When it is ambiguous whether we should configure for Console/a child Process,
+// we prefer configuring for the Console.
+static bool g_reading = false;                // tracks whether the application is performing a Console.Read operation
+static bool g_childUsesTerminal = false;      // tracks whether a child process is using the terminal
+static bool g_terminalUninitialized = false;  // tracks whether the application is terminating
+
+static bool g_noTty = false;                  // cache we are not a tty
+
+static volatile bool g_receivedSigTtou = false;
+
+static void ttou_handler(int signo)
 {
-    // pal_signal.cpp calls this on SIGQUIT/SIGINT.
-    // This can happen when SystemNative_InitializeConsole was not called.
-
-    // Put the attributes back to what they were when the console was initially initialized.
-    // We only do so, however, if we have explicitly modified the termios; doing so always
-    // can result in problems if the app is in the background, as then attempting to call
-    // tcsetattr on STDIN_FILENO will suspend the app and prevent its shutdown. We also don't
-    // want to, for example, just compare g_currTermios with g_initTermios, as we'd then be
-    // factoring in changes made by other apps or by user code.
-    if (g_haveInitTermios &&                     // we successfully initialized the console
-        (g_readInProgress || !g_signalForBreak)) // we modified attributes
+    (void)signo;
+    g_receivedSigTtou = true;
+}
+
+static void InstallTTOUHandler(void (*handler)(int))
+{
+    struct sigaction action;
+    memset(&action, 0, sizeof(action));
+    action.sa_handler = handler;
+    int rvSigaction = sigaction(SIGTTOU, &action, NULL);
+    assert(rvSigaction == 0);
+    (void)rvSigaction;
+}
+
+static bool TcSetAttr(struct termios* termios, bool blockIfBackground)
+{
+    if (g_terminalUninitialized)
     {
-        tcsetattr(STDIN_FILENO, TCSANOW, &g_initTermios);
-        // ignore any failure
+        // The application is exiting, we mustn't change terminal settings.
+        return true;
     }
+
+    if (!blockIfBackground)
+    {
+        // When the process is running in background, changing terminal settings
+        // will stop it (default SIGTTOU action).
+        // We change SIGTTOU to to get EINTR instead of blocking.
+        InstallTTOUHandler(ttou_handler);
+
+        g_receivedSigTtou = false;
+    }
+
+    bool rv = tcsetattr(STDIN_FILENO, TCSANOW, termios) >= 0;
+
+    if (!blockIfBackground)
+    {
+        if (!rv && errno == EINTR && g_receivedSigTtou)
+        {
+            // Operation failed because we are background
+            // pretend it went fine.
+            rv = true;
+        }
+
+        // Restore default SIGTTOU handler.
+        InstallTTOUHandler(SIG_DFL);
```

Comment:
Does any caller exam errno, and if yes, might this change it?

Reply:
Because InstallTTOUHandler is assumed to be always successful, it shouldn't change errno.